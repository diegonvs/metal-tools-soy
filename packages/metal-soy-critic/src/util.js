const chalk = require('chalk');
const Promise = require('bluebird');

function combineResults(first, second) {
  return toResult(
    first.status && second.status,
    ...first.messages.concat(second.messages)
  );
}

function difference(setA, setB) {
  const difference = new Set();
  for (const elem of setA) {
    if (!setB.has(elem)) {
      difference.add(elem);
    }
  }
  return difference;
}

function toResult(status, ...messages) {
  return {
    messages,
    status
  };
}

function parseTemplateName(rawName) {
  const segments = rawName.split('.');
  const namespace = segments
    .slice(0, segments.length - 1)
    .join('.');

  return {
    name: segments[segments.length - 1],
    namespace: namespace || null
  };
}

function sequence(...tasks) {
  return Promise.reduce(tasks, (res, next) =>
    next().then(value => [...res, value]),
    []
  );
}

function joinErrors(lines) {
  return lines
    .map(line => chalk.red(line))
    .join('\n');
}

module.exports = {
  combineResults,
  difference,
  joinErrors,
  parseTemplateName,
  sequence,
  toResult
};
