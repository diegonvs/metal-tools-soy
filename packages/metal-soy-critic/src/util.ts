import * as chalk from 'chalk';

export interface Result {
  status: boolean;
  messages: Array<string>;
}

export function combineResults(first: Result, second: Result): Result {
  return toResult(
    first.status && second.status,
    ...first.messages.concat(second.messages)
  );
}

export function toResult(status: boolean, ...messages: Array<string>): Result {
  return {
    messages,
    status
  };
}

export interface TemplateName {
  name: string,
  namespace: string | null
}

export function difference<T>(setA: Set<T>, setB: Set<T>): Set<T> {
  const difference = new Set();
  for (const elem of setA) {
    if (!setB.has(elem)) {
      difference.add(elem);
    }
  }
  return difference;
}

export function joinErrors(lines: Array<string>): string {
  return lines
    .map(line => chalk.red(line))
    .join('\n');
}

export function isSorted(items: Array<string>): boolean {
  const sortedItems = items
    .concat()
    .sort((a, b) => a.localeCompare(b));

  return sortedItems.join() === items.join();
}
