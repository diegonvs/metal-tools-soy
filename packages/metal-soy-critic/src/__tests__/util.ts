import * as Util from '../util';

describe('toResult', () => {
  test('should return a Result object', () => {
    const result = Util.toResult(true, 'message');

    expect(result).toEqual({
      status: true,
      messages: ['message']
    });
  });

  test('should default to empty messages', () => {
    const result = Util.toResult(false);

    expect(result).toEqual({
      status: false,
      messages: []
    });
  });
});

describe('joinErrors', () => {
  test('should return a string with all messages joined', () => {
    const errors = ['nope', 'bad', 'try again'];

    expect(Util.joinErrors(errors)).toMatchSnapshot();
  });
});

describe('combineResults', () => {
  test('should combine two results into a single result', () => {
    const result = Util.combineResults(
      {
        messages: [],
        status: true
      },
      {
        messages: [],
        status: true
      }
    );

    expect(result).toEqual({
      status: true,
      messages: []
    });
  });

  test('if one fails, they result fails', () => {
    const result = Util.combineResults(
      {
        messages: ['nope'],
        status: false
      },
      {
        messages: [],
        status: true
      }
    );

    expect(result).toEqual({
      status: false,
      messages: ['nope']
    });
  });

  test('should combine messages', () => {
    const result = Util.combineResults(
      {
        messages: ['nope'],
        status: false
      },
      {
        messages: ['nah'],
        status: false
      }
    );

    expect(result).toEqual({
      status: false,
      messages: ['nope', 'nah']
    });
  });
});

describe('difference', () => {
  test('should return the difference of two sets', () => {
    const setA = new Set([1, 2, 3]);
    const setB = new Set([1, 2, 4]);

    expect(Util.difference(setA, setB)).toEqual(new Set([3]));
    expect(Util.difference(setB, setA)).toEqual(new Set([4]));
  })
})

describe('parseTemplateName', () => {
  test('should parse template name', () => {
    const result = Util.parseTemplateName('.render');

    expect(result).toMatchObject({
      name: 'render',
      namespace: null
    });
  })

  test('should parse a namespace', () => {
    const result = Util.parseTemplateName('MyNamespace.render');

    expect(result).toMatchObject({
      name: 'render',
      namespace: 'MyNamespace'
    });
  })

  test('should parse nested namespaces', () => {
    const result = Util.parseTemplateName('My.Nested.Namespace.render');

    expect(result).toMatchObject({
      name: 'render',
      namespace: 'My.Nested.Namespace'
    });
  })
});

describe('reverseJoin', () => {
  test('should combine two elements', () => {
    const result = Util.reverseJoin(1, 2);

    expect(result).toEqual([1, 2]);
  });

  test('should combine an element and an array of elements', () => {
    const result = Util.reverseJoin(1, [2, 3, 4]);

    expect(result).toEqual([1, 2, 3, 4,]);
  });
});

describe('isSorted', () => {
  test('should return false', () => {
    const word = ['one', 'two', 'three'];

    expect(Util.isSorted(word)).toBe(false);
  });

  test('should return true', () => {
    const word = ['one', 'three', 'two'];

    expect(Util.isSorted(word)).toBe(true);
  });
});
