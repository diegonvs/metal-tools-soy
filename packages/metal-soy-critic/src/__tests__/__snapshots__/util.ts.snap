// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`joinErrors should return a string with all messages joined 1`] = `
"[31mnope[39m
[31mbad[39m
[31mtry again[39m"
`;
